pkg/apis/allocation/v1/gameserverallocation.go

```diff
@@ -82,6 +82,137 @@ type GameServerAllocationSpec struct {
+	MaxAvailable int64 `json:"maxAvailable,omitempty"`
+}
+
+// ApplyDefaults applies default values to the PlayerSelector
+func (s *GameServerSelector) ApplyDefaults() {
+	if runtime.FeatureEnabled(runtime.FeatureStateAllocationFilter) {
+		if s.GameServerState == nil {
+			state := agonesv1.GameServerStateReady
+			s.GameServerState = &state
+		}
+	}
+
+	if runtime.FeatureEnabled(runtime.FeaturePlayerAllocationFilter) {
+		if s.Players == nil {
+			s.Players = &PlayerSelector{}
+		}
+	}
+}
+
+// Matches checks to see if a GameServer matches a given GameServerSelector's criteria
+func (s *GameServerSelector) Matches(gs *agonesv1.GameServer) bool {
+
+	// Assume at this point, this has already been run through Validate(), and it can be converted.
+	// We end up running LabelSelectorAsSelector twice for each allocation, but if we store the results of this
+	// function within the GameServerSelector, we can't fuzz the GameServerAllocation as reflect.DeepEqual
+	// will fail due to the unexported field.
+	selector, _ := metav1.LabelSelectorAsSelector(&s.LabelSelector) // nolint: error
+	if selector == nil {
+		return false
+	}
+
+	// first check labels
+	if !selector.Matches(labels.Set(gs.ObjectMeta.Labels)) {
+		return false
+	}
+
+	// then if state is being checked, check state
+	if runtime.FeatureEnabled(runtime.FeatureStateAllocationFilter) {
+		if s.GameServerState != nil && gs.Status.State != *s.GameServerState {
+			return false
+		}
+	}
+
+	// then if player count is being checked, check that
+	if runtime.FeatureEnabled(runtime.FeaturePlayerAllocationFilter) {
+		// 0 is unlimited number of players
+		if s.Players != nil && gs.Status.Players != nil && s.Players.MaxAvailable != 0 {
+			available := gs.Status.Players.Capacity - gs.Status.Players.Count
+			if !(available >= s.Players.MinAvailable && available <= s.Players.MaxAvailable) {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+// Validate validates that the selection fields have valid values
+func (s *GameServerSelector) Validate(field string, causes []metav1.StatusCause) ([]metav1.StatusCause, bool) {
+	valid := true
+
+	_, err := metav1.LabelSelectorAsSelector(&s.LabelSelector)
+	if err != nil {
+		valid = false
+		causes = append(causes, metav1.StatusCause{
+			Type:    metav1.CauseTypeFieldValueInvalid,
+			Message: fmt.Sprintf("Error converting label selector: %s", err),
+			Field:   field,
+		})
+	}
+
+	if runtime.FeatureEnabled(runtime.FeatureStateAllocationFilter) {
+		if s.GameServerState != nil && !(*s.GameServerState == agonesv1.GameServerStateAllocated || *s.GameServerState == agonesv1.GameServerStateReady) {
+			valid = false
+			causes = append(causes, metav1.StatusCause{
+				Type:    metav1.CauseTypeFieldValueInvalid,
+				Message: "GameServerState value can only be Allocated or Ready",
+				Field:   field,
+			})
+		}
+	}
+
+	if runtime.FeatureEnabled(runtime.FeaturePlayerAllocationFilter) && s.Players != nil {
+		if s.Players.MinAvailable < 0 {
```

Comment:
Is there any case where we should check for the upper bound? Or any valid int value is acceptable?

Reply:
Not that I can think of. The extra thing is `if s.Players.MinAvailable > s.Players.MaxAvailable` to make sure min is not greater than max.